src/sql/sql_provider_datafusion/mod.rs

```diff
@@ -76,38 +98,37 @@ impl<T, P> SqlTable<T, P> {
             .await
             .context(UnableToGetSchemaSnafu)?;
 
-        Ok(Self {
+        Ok(Self::new_with_schema(
             name,
-            pool: Arc::clone(pool),
+            pool,
             schema,
             table_reference,
             engine,
-            dialect: None,
-        })
+            dialect,
+        ))
     }
 
     pub fn new_with_schema(
         name: &'static str,
         pool: &Arc<dyn DbConnectionPool<T, P> + Send + Sync>,
         schema: impl Into<SchemaRef>,
         table_reference: impl Into<TableReference>,
-        engine: Option<expr::Engine>,
+        engine: Option<Engine>,
+        dialect: Option<Arc<dyn Dialect + Send + Sync>>,
     ) -> Self {
+        let dialect = dialect.unwrap_or(
+            engine
+                .and_then(|e| e.dialect())
+                .unwrap_or(Arc::new(DefaultDialect {})),
+        );
+
         Self {
             name,
             pool: Arc::clone(pool),
             schema: schema.into(),
             table_reference: table_reference.into(),
             engine,
-            dialect: None,
-        }
-    }
-
-    #[must_use]
-    pub fn with_dialect(self, dialect: Arc<dyn Dialect + Send + Sync>) -> Self {
-        Self {
-            dialect: Some(dialect),
-            ..self
+            dialect,
```

Comment:
I would prefer to keep the `with_dialect` function. This makes it cleaner to override the default dialect if required without cluttering the main constructor. The main constructor can now lose two arguments (for engine and dialect).